File: app/eta.rb
class ETA
  attr_reader :arrival

  def <=>(other)
    @arrival.<=>(other.arrival)
  end

  def initialize
    @destination = ''
    @route = ''
    @approaching = ''
    @arrival = ''
    @generated = ''
  end

  def dest_append(string)
    @destination << string
  end

  def rt_append(string)
    @route << string
  end

  def app_append(string)
    @approaching << string
  end

  def arr_append(string)
    @arrival << string
  end

  def prd_append(string)
    @generated << string
  end

  def destination
    @destination
  end

  def route
    CTAInfo.routes.has_key?(@route) ? CTAInfo.routes[@route] : @route
  end

  def time
    @@formatter ||= begin
      formatter = NSDateFormatter.alloc.init
      formatter.dateFormat = 'yyyyMMdd HH:mm:ss'
      formatter.timeZone = NSTimeZone.localTimeZone
      formatter
    end

    @time ||= begin
      if @approaching == '1'
        0
      else
        arrival = @@formatter.dateFromString(@arrival)
        generated = @@formatter.dateFromString(@generated)
        arrival - generated
      end
    end
  end

  def time_string
    time_min = (time / 60.0).floor
    return 'Due' if @approaching == '1' || time_min < 1
    "#{time_min} min#{time_min == 1 ? '' : 's'}"
  end
end

File: Rail Line/ETA.swift
import Foundation

class ETA {
    private var destination = ""
    private var route = ""
    private var approaching = ""
    private var arrival = ""
    private var generated = ""
    private var time: Double?
    private static let formatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyyMMdd HH:mm:ss"
        formatter.timeZone = NSTimeZone.local
        return formatter
    }()

    func destAppend(_ val: String) {
        destination.append(contentsOf: val)

        if destination == "O'Hare" || destination == "Midway" {
            destination.append(contentsOf: " ✈️")
        }
    }

    func rtAppend(_ val: String) {
        route.append(contentsOf: val)
    }
    func appAppend(_ val: String) {
        approaching.append(contentsOf: val)
    }

    func arrAppend(_ val: String) {
        arrival.append(contentsOf: val)
    }

    func prdAppend(_ val: String) {
        generated.append(contentsOf: val)
    }

    func getRoute() -> String {
        return Routes.mapCode(route)
    }

    func getDestination() -> String {
        return destination
    }

    func getTime() -> Double {
        if time == nil {
            if approaching == "1" {
                time = 0.0
            } else {
                let arrivalDate = ETA.formatter.date(from: arrival)
                let generatedDate = ETA.formatter.date(from: generated)
                time = arrivalDate!.timeIntervalSince(generatedDate!)
            }
        }
        return time!
    }

    func getTimeString() -> String {
        let minutes = Int(getTime() / 60.0)
        if approaching == "1" || minutes < 1 { return "Due" }
        return "\(minutes) min\(minutes == 1 ? "" : "s")"
    }
}

func < (lhs: ETA, rhs: ETA) -> Bool {
    return lhs.getTime() < rhs.getTime()
}

func == (lhs: ETA, rhs: ETA) -> Bool {
    return lhs.getTime() == rhs.getTime()
}
